src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1013,31 +1021,69 @@ public SolutionCompilationState WithOptions(SolutionOptionSet options)
             this.SolutionState.WithOptions(options));
     }
 
-    public Task<SolutionCompilationState> WithFrozenPartialCompilationsAsync(CancellationToken cancellationToken)
-        => _cachedFrozenSnapshot.GetValueAsync(cancellationToken);
+    public SolutionCompilationState WithFrozenPartialCompilations(CancellationToken cancellationToken)
+    {
+        using (this.StateLock.DisposableWait(cancellationToken))
+        {
+            _cachedFrozenSnapshot ??= ComputeFrozenSnapshot(cancellationToken);
+            return _cachedFrozenSnapshot;
+        }
+    }
 
     private SolutionCompilationState ComputeFrozenSnapshot(CancellationToken cancellationToken)
     {
         var newIdToProjectStateMapBuilder = this.SolutionState.ProjectStates.ToBuilder();
         var newIdToTrackerMapBuilder = _projectIdToTrackerMap.ToBuilder();
 
+        using var _1 = ArrayBuilder<DocumentState>.GetInstance(out var documentsToRemove);
+        using var _2 = ArrayBuilder<DocumentState>.GetInstance(out var documentsToAdd);
+
         foreach (var projectId in this.SolutionState.ProjectIds)
         {
             // if we don't have one or it is stale, create a new partial solution
-            var tracker = GetCompilationTracker(projectId);
-            var newTracker = tracker.FreezePartialState(cancellationToken);
+            var oldTracker = GetCompilationTracker(projectId);
+            var newTracker = oldTracker.FreezePartialState(cancellationToken);
+            if (oldTracker == newTracker)
+                continue;
 
             Contract.ThrowIfFalse(newIdToProjectStateMapBuilder.ContainsKey(projectId));
-            newIdToProjectStateMapBuilder[projectId] = newTracker.ProjectState;
+
+            var oldProjectState = this.SolutionState.GetRequiredProjectState(projectId);
+            var newProjectState = newTracker.ProjectState;
+
+            newIdToProjectStateMapBuilder[projectId] = newProjectState;
             newIdToTrackerMapBuilder[projectId] = newTracker;
+
+            // Freezing projects can cause them to have an entirely different set of documents (since it effectively
+            // rewinds the project back to the last time it produced a compilation).  Ensure we keep track of the docs
+            // added or removed from the project states to keep the final filepath-to-documentid map accurate.
+
+            foreach (var (documentId, oldDocumentState) in oldProjectState.DocumentStates.States)
+            {
+                if (!newProjectState.DocumentStates.States.ContainsKey(documentId))
+                    documentsToRemove.Add(oldDocumentState);
+            }
+
+            foreach (var (documentId, newDocumentState) in newProjectState.DocumentStates.States)
+            {
+                if (!oldProjectState.DocumentStates.States.ContainsKey(documentId))
+                    documentsToAdd.Add(newDocumentState);
+            }
```

Comment:
@jasonmalinowski for if you think there's an easier or more efficient way to do this.

Reply:
We have https://github.com/dotnet/roslyn/blob/f656544b1a42c3d212362dab79829e0b3abe30f4/src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs#L195-L202 that can help you out here.

Reply:
Although I realize you'll need diff states, not just the IDs, since our APIs at least would allow a document's file path to change without changing the ID.

Reply:
switched to those helpers.